Refactor the internal sql interface somewhat

flub opened the pull request:
Experiment with refactoring the internal sql interface a bit.  My
original goal was to modify the schema and thus refactor to a state
where it would be sane to write tests for dc_sqlite_open() (and/or
however it ends up refactored) to assert schame changes before/after.

flub commented:
This is much less than i planned to do but probably a good place to stop since #142 is also in progress and carrying on now would keep introducing more conflicts.

Includes #144 wholly so would obsolete that PR is merged.  If that one is merged first this will need to resolve some conflicts.

r10s commented on src/dc_chat.rs:
```diff
@@ -273,7 +273,7 @@ pub unsafe fn dc_create_or_lookup_nchat_by_contact_id(
         *ret_chat_blocked = 0i32
     }
-    if context.sql.read().unwrap().cobj.is_null() {
+    if context.sql.is_open() {
```
i think, this should be `!is_open()`

r10s commented on src/dc_chat.rs:
```diff
@@ -363,13 +363,13 @@ pub unsafe fn dc_lookup_real_nchat_by_contact_id(
         *ret_chat_blocked = 0i32
     }
-    if context.sql.clone().read().unwrap().cobj.is_null() {
+    if context.sql.is_open() {
```
i think, this should be `!is_open()`

r10s commented on src/dc_contact.rs:
```diff
@@ -1219,10 +1208,10 @@ pub unsafe fn dc_get_contact_origin(
     let mut stmt: *mut sqlite3_stmt = 0 as *mut sqlite3_stmt;
     let mut ret = false;
-    if !(context.sql.clone().read().unwrap().cobj.is_null() || contact_id <= 9i32 as libc::c_uint) {
+    if !(context.sql.is_open() || contact_id <= 9i32 as libc::c_uint) {
```
i think this should be `if !(!context.sql.is_open()...`

r10s commented on src/dc_msg.rs:
```diff
@@ -1420,17 +1416,17 @@ pub unsafe fn dc_mdn_from_ext(
     let mut stmt: *mut sqlite3_stmt = 0 as *mut sqlite3_stmt;
     let mut ret: size_t = 0i32 as size_t;
-    if !(*&context.sql.clone().read().unwrap()).cobj.is_null() {
+    if !(*&context.sql).is_open() {
```
i think this should be `if (*&context.sql).is_open()`

r10s commented on src/dc_sqlite3.rs:
```diff
@@ -13,43 +13,67 @@ const DC_OPEN_READONLY: usize = 0x01;
-    if !sql.cobj.is_null() {
-        dc_sqlite3_close(context, sql);
+    pub fn is_open(&self) -> bool {
+        !self.cobj.read().unwrap().is_null()
```
i do not understand this line - where did the close() go to?

flub replied:
the dc_sqlite3_unref() function is completely gone.  is_open() should not close the connection... and replaces the dc_sqlite3_is_open() call (which is also gone).  The diff is just confusing putting the wrong things next to each other.

flub commented:
So the segfault happens in dc_log_info because that is the only test in which the callback of the context is a real C NULL.  Why this didn't segfault before i have no idea.

flub commented:
I've done a very careful review of the diff here and tweaked a few things minorly: removed some changes that shouldn't yet be made (was altering the sql schema - oops) and removed some small cleanup diffs that didn't have anything to do with this cl.

Since #153 is merged this works on CI and that was an edge case which doesn't need @r10s' tests redone.  So I'm going to merge this now once the CI finishes again.
